src/lib/mina_lib/mina_lib.ml

```diff
@@ -1495,6 +1495,138 @@ let start_filtered_log ~commit_id
+  let open Network_peer in
+  let%bind all_peers = Mina_networking.peers network in
+  let peer_limit = 5 in
+  let limited_peers = List.take all_peers peer_limit in
+  Deferred.List.iter
+    ~f:(fun peer ->
+      [%log debug] "PEER IS: Fetching completed snarks from peer: $peer"
+        ~metadata:[ ("peer", Network_peer.Peer.to_yojson peer) ] ;
+      let completed_works =
+        Mina_networking.get_completed_checked_snarks network peer
+      in
+      let%bind completed_works = completed_works in
+      let completed_works =
+        match completed_works with
+        | Error e ->
+            [%log debug]
+              ~metadata:
+                [ ("peer", Network_peer.Peer.to_yojson peer)
+                ; ("error", Error_json.error_to_yojson e)
+                ]
+              "Failed to fetch completed snarks from peer: $error" ;
+            []
+        | Ok completed_works ->
+            completed_works
+      in
+
+      [%log debug]
+        ~metadata:
+          [ ("peer", Network_peer.Peer.to_yojson peer)
+          ; ("completed_works", `Int (List.length completed_works))
+          ]
+        "Fetched $completed_works completed snarks from peer: $peer" ;
+
+      (* Keep reading from the transition frontier until it has caught up to the most valid block from the network.
+         * This is to ensure that the snarks are verified and added to the pool in the correct order
+      *)
+      let rec wait_for_new_top_block received_block =
+        let frontier = get_current_frontier () in
+        match frontier with
+        | None ->
+            [%log error]
+              "Transition frontier is not available after sync something has \
+               gone terribly wrong" ;
+            let%bind () = after (Time.Span.of_ms 20.) in
+            wait_for_new_top_block received_block
+        | Some frontier ->
+            let tip = Transition_frontier.best_tip frontier in
+            let top_block =
+              Transition_frontier.Breadcrumb.validated_transition tip
+              |> Mina_block.Validated.header
+              |> Mina_block.Header.blockchain_length
+            in
+            [%log debug]
+              ~metadata:
+                [ ( "old_top_block"
+                  , `Int (received_block |> Unsigned.UInt32.to_int) )
+                ; ("new_top_block", `Int (top_block |> Unsigned.UInt32.to_int))
+                ]
+              "WAITING  old top block: $old_top_block, new top block: \
+               $new_top_block" ;
+            let delta =
+              Unsigned.UInt32.(Infix.(received_block - top_block) |> to_int)
+            in
+            (* if delta is less than or equal to zero the transition frontier has caught up with the network *)
+            if delta <= 0 then Deferred.unit
+            else
+              let%bind () = after (Time.Span.of_ms 20.) in
+              wait_for_new_top_block received_block
+      in
+
+      let%bind () = wait_for_new_top_block received_block in
```

Comment:
Given the current implementation, this can be moved outside the loop.

Reply:
done